drivers/spi/spi_pl022.c

```diff
@@ -0,0 +1,538 @@
+			       const struct spi_config *spicfg)
+{
+	const struct spi_pl022_cfg *cfg = dev->config;
+	const uint16_t op = spicfg->operation;
+	struct spi_pl022_data *data = dev->data;
+	uint32_t cr0;
+	uint32_t cr1;
+	uint32_t prescale;
+	uint32_t postdiv;
+
+	if (spi_context_configured(&data->ctx, spicfg)) {
+		return 0;
+	}
+
+	if (spicfg->frequency > MAX_FREQ_CONTROLLER_MODE(cfg)) {
+		LOG_ERR("Frequency is up to %u in controller mode.", MAX_FREQ_CONTROLLER_MODE(cfg));
+		return -ENOTSUP;
+	}
+
+	if (op & SPI_TRANSFER_LSB) {
+		LOG_ERR("LSB-first not supported");
+		return -ENOTSUP;
+	}
+
+	/* Half-duplex mode has not been implemented */
+	if (op & SPI_HALF_DUPLEX) {
+		LOG_ERR("Half-duplex not supported");
+		return -ENOTSUP;
+	}
+
+	/* Peripheral mode has not been implemented */
+	if (SPI_OP_MODE_GET(op) != SPI_OP_MODE_MASTER) {
+		LOG_ERR("Peripheral mode is not supported");
+		return -ENOTSUP;
+	}
+
+	/* Word sizes other than 8 bits has not been implemented */
+	if (SPI_WORD_SIZE_GET(op) != 8) {
+		LOG_ERR("Word sizes other than 8 bits are not supported");
+		return -ENOTSUP;
+	}
+
+	/* configure registers */
+
+	prescale = spi_pl022_calc_prescale(cfg->pclk, spicfg->frequency);
+	postdiv = spi_pl022_calc_postdiv(cfg->pclk, spicfg->frequency, prescale);
+
+	cr0 = 0;
+	cr0 |= (postdiv << SSP_CR0_SCR_POS);
+	cr0 |= (SPI_WORD_SIZE_GET(op) - 1);
+	cr0 |= (op & SPI_MODE_CPOL) ? SSP_CR0_MASK_SPO : 0;
+	cr0 |= (op & SPI_MODE_CPHA) ? SSP_CR0_MASK_SPH : 0;
+
+	cr1 = 0;
+	cr1 |= SSP_CR1_MASK_SSE; /* Always enable SPI */
+	cr1 |= (op & SPI_MODE_LOOP) ? SSP_CR1_MASK_LBM : 0;
+
+	SSP_WRITE_REG(SSP_CPSR(cfg->reg), prescale);
+	SSP_WRITE_REG(SSP_CR0(cfg->reg), cr0);
+	SSP_WRITE_REG(SSP_CR1(cfg->reg), cr1);
+
+	data->ctx.config = spicfg;
+
+	return 0;
+}
+
+/*
+ * calculate actual transfer length.
+ * If one buffer is invalid, use length of others,
+ * IF both buffer is valid, using length of shorter one.
+ */
+static inline size_t spi_pl022_xfer_length(const void *txbuf, const size_t txlen,
+					   void *rxbuf, const size_t rxlen)
+{
+	size_t min = (txlen < rxlen ? txlen : rxlen);
+
+	if (!txbuf && txlen == 0) {
+		return rxlen;
+	} else if (!rxbuf && rxlen == 0) {
+		return txlen;
+	}
+	return min;
+}
+
+static size_t spi_pl022_xfer(const struct device *dev, const void *txbuf, const size_t txlen,
+			     void *rxbuf, const size_t rxlen)
+{
+	const struct spi_pl022_cfg *cfg = dev->config;
+	const size_t xferlen = spi_pl022_xfer_length(txbuf, txlen, rxbuf, rxlen);
+	size_t rx_remaining = xferlen;
+	size_t tx_remaining = xferlen;
+
+	/* Ensure writable */
+	while (!SSP_TX_FIFO_EMPTY(cfg->reg))
+		;
+	/* Drain RX FIFO */
+	while (SSP_RX_FIFO_NOT_EMPTY(cfg->reg))
+		SSP_READ_REG(SSP_DR(cfg->reg));
+
+	while (rx_remaining > 0 || tx_remaining > 0) {
+		/* Fill up fifo with available TX data */
+		while (SSP_TX_FIFO_NOT_FULL(cfg->reg) && tx_remaining > 0) {
+			/* Send 0 in the case of read operation */
+			uint32_t txd = 0;
```

Comment:
Declare those variables at the beginning of the function

Reply:
Fixed it.